File: hooks/scripts/artifact-gate.mjs
#!/usr/bin/env node
// artifact-gate.mjs — PreToolUse hook for pipeline_complete_phase
// Verifies at least one artifact exists for the target phase before allowing completion

import { readFileSync, readdirSync, existsSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const RUNS_DIR = join(__dirname, '..', '..', 'pipeline_mcp_data', 'runs')

function findLatestRunState() {
  if (!existsSync(RUNS_DIR)) return null

  const runDirs = readdirSync(RUNS_DIR, { withFileTypes: true })
    .filter(d => d.isDirectory())
    .map(d => d.name)

  let latestRun = null
  let latestTime = 0

  for (const dir of runDirs) {
    const statePath = join(RUNS_DIR, dir, 'run-state.json')
    if (!existsSync(statePath)) continue
    try {
      const state = JSON.parse(readFileSync(statePath, 'utf-8'))
      const updatedAt = new Date(state.updated_at).getTime()
      if (updatedAt > latestTime) {
        latestTime = updatedAt
        latestRun = state
      }
    } catch {
      continue
    }
  }

  return latestRun
}

let input = ''
for await (const chunk of process.stdin) {
  input += chunk
}

try {
  const hookInput = JSON.parse(input || '{}')
  const toolInput = hookInput.tool_input || {}
  const phaseName = toolInput.phase || toolInput.phase_name

  if (!phaseName) {
    process.stdout.write(JSON.stringify({ permissionDecision: 'allow' }))
    process.exit(0)
  }

  const runState = findLatestRunState()

  if (!runState) {
    process.stdout.write(JSON.stringify({ permissionDecision: 'allow' }))
    process.exit(0)
  }

  // Check available_artifacts for this phase
  const phaseArtifacts = (runState.available_artifacts || [])
    .filter(a => a.phase === phaseName)

  // Also check phase output_artifacts
  const phaseState = runState.phases[phaseName]
  const outputArtifacts = phaseState?.output_artifacts || []

  if (phaseArtifacts.length === 0 && outputArtifacts.length === 0) {
    process.stdout.write(JSON.stringify({
      permissionDecision: 'deny',
      deny_reason: `Cannot complete phase "${phaseName}" — no artifacts have been stored. Use pipeline_store_artifact to save at least one output before completing.`,
    }))
  } else {
    process.stdout.write(JSON.stringify({ permissionDecision: 'allow' }))
  }
} catch {
  process.stdout.write(JSON.stringify({ permissionDecision: 'allow' }))
}

File: hooks/scripts/artifact_gate.py
#!/usr/bin/env python3
"""Verify a phase has artifacts before allowing completion — PreToolUse hook."""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from hooks.lib.common import find_latest_run_state, read_stdin  # noqa: E402

_ROOT = Path(__file__).resolve().parents[2]
RUNS_DIR = (
    Path(os.environ.get("PIPELINE_MCP_DATA_DIR") or (_ROOT / "pipeline_mcp_data"))
    / "runs"
)


def main() -> None:
    raw_input = read_stdin()

    try:
        hook_input: Any = json.loads(raw_input or "{}")
        tool_input: dict[str, Any] = hook_input.get("tool_input") or {}
        phase_name = tool_input.get("phase") or tool_input.get("phase_name")

        if not phase_name:
            sys.stdout.write(json.dumps({"permissionDecision": "allow"}))
            sys.exit(0)

        run_state = find_latest_run_state(RUNS_DIR)

        if not run_state:
            sys.stdout.write(json.dumps({"permissionDecision": "allow"}))
            sys.exit(0)

        # Check available_artifacts for this phase
        phase_artifacts = [
            a
            for a in (run_state.get("available_artifacts") or [])
            if a.get("phase") == phase_name
        ]

        # Also check phase output_artifacts
        phases: dict[str, Any] = run_state["phases"]
        phase_state: dict[str, Any] | None = phases.get(phase_name)
        output_artifacts = (
            (phase_state.get("output_artifacts") or []) if phase_state else []
        )

        if len(phase_artifacts) == 0 and len(output_artifacts) == 0:
            sys.stdout.write(
                json.dumps(
                    {
                        "permissionDecision": "deny",
                        "deny_reason": (
                            f'Cannot complete phase "{phase_name}" — no '
                            f"artifacts have been stored. Use "
                            f"pipeline_store_artifact to save at least one "
                            f"output before completing."
                        ),
                    }
                )
            )
        else:
            sys.stdout.write(json.dumps({"permissionDecision": "allow"}))
    except Exception:
        sys.stdout.write(json.dumps({"permissionDecision": "allow"}))


if __name__ == "__main__":
    main()
